0011-B-Distributed-Lock-building-block.md

```diff
@@ -0,0 +1,359 @@
+ 	TTL time.Duration `json:"ttl"`
+}
+
+// LockResponse is the response used by Lock and TryLock when the operation is completed.
+type LockResponse struct {
+ 	// ID of the lock.
+ 	LockID  string `json:"lockID"`
+}
+
+// RenewLockRequest is a lock renewal request.
+type RenewLockRequest struct {
+ 	// ID of the resource whose lock needs to be renewed.
+ 	// This is an arbitrary value supplied by the user.
+ 	ResourceID string `json:"resourceId"`
+ 	// ID of the lock.
+ 	LockID any `json:"lockID"`
+ 	// Maximum duration for the lock (if not renewed).
+ 	// If <= 0, the lock uses the default TTL set in the component.
+ 	TTL time.Duration `json:"ttl"`
+}
+
+// UnlockRequest is a lock release request.
+type UnlockRequest struct {
+ 	// ID of the resource that is locked.
+ 	ResourceID string `json:"resourceId"`
+ 	// ID of the lock.
+ 	LockID  string `json:"lockID"`
+ 	// If true, will force the lock to be released, even if the lockID does not match.
+ 	// This may not be supported by all lock stores.
+ 	// This is a potentially destructive operation that should be used with caution.
+ 	Force bool `json:"force"`
+}
+```
+
+The main differences include:
+
+- New blocking `Lock` and `RenewLock` methods.
+- Removed the concept of "lock owners", supplied by the caller, in favor of lock IDs that are generated by the component.
+- Standard errors are returned to indicate situations such as resource already locked or resources that fails to be unlocked (or whose lock fails to be renewed) due to an invalid lock ID.
+- TTLs are pased as `time.Duration`, and components have a default TTL.
+- Added a `force` option in `Unlock`.
+
+### Runtime APIs
+
+The new APIs aim at offering a simplified user experience where developers interact with at most 3 Dapr APIs, and Dapr takes care of all background activities including renewing locks. This allows Dapr to provide additional, differentiated value to users too.
+
+We propose two possible ways to implement the new APIs, which are both proposed below, each with pros and cons. The difference lies in how the user removes the lock.
+
+**In both cases**:
+
+- Dapr renews locks automatically in background. Apps do not need to invoke a method to renew a lock explicitly. The TTL sent in requests is only used to control how frequently locks are to be renewed (lower TTLs cause locks to be renewed more frequently, but allow for quicker detection of failed hosts).
+- If the app becomes unhealthy (when app healthchecks are enabled) or if the sidecar crashes, locks are not renewed and are left to expire when the TTL ends.
+
+> About force unlocking: this is implemented primarily to recover from situations where the lock was owned by an app that is known to have crashed. This is a potentially dangerous activity, as the previous owner of the lock is not informed (at least, not immediately) that the lock has been taken away, at least not until it tries to renew or relinquish the lock.  
```

Comment:
I'm quite torn as to whether we should include this. I can see the use case, however the ramifications of two apps thinking they both have a lock could potentially be huge and is giving developers a big foot gun. Perhaps we hold off on exposing this API and reevaluating if it comes up as a feature request? Not being implemented by all components is also a kicker.

Reply:
I have been on the fence myself the entire time about this.

Proposals often indicate the "desired state", and they don't have to be implemented in full at the beginning. We can always create the API in the component and then expose it later.